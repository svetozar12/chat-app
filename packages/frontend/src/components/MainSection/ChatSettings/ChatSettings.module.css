.removeUser {
  width: 2rem;
  height: 2rem;
  cursor: pointer;
}

.removeUser:hover {
  color: red !important;
}
