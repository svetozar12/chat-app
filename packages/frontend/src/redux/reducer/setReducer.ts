import { Action, ActionType, InitialState2 } from "../state";
export const initialState = {
  reciever: "",
  chat_inviter: "",
  pageNumber: 2,
  setUserSettings: false,
  setFriendRequest: false,
  toggleCreateGroup: false,
  setMobileNav: false,
  setIsMatch: false,
  setChatSettings: false,
  setModalInvite: false,
  setIsLoggingIn: false,
};
const setReducer = (state: InitialState2 | any | InitialState2[] = initialState, action: Action) => {
  switch (action.type) {
    case ActionType.SET_RECIEVER:
      return { ...state, reciever: action.payload };
    case ActionType.CHAT_INVITER:
      return { ...state, chat_inviter: action.payload };
    case ActionType.INCREMENT_PAGE_NUMBER:
      return { ...state, pageNumber: ++action.payload };
    case ActionType.SET_FRIEND_REQUEST:
      return { ...state, setFriendRequest: action.payload };
    case ActionType.SET_USER_SETTINGS:
      return { ...state, setUserSettings: action.payload };
    case ActionType.TOGGLE_CREATE_GROUP:
      return { ...state, toggleCreateGroup: action.payload };
    case ActionType.SET_MOBILE_NAV:
      return { ...state, setMobileNav: action.payload };
    case ActionType.SET_CHAT_SETTINGS:
      return { ...state, setChatSettings: action.payload };
    case ActionType.SET_IS_MATCH:
      return { ...state, setIsMatch: action.payload };
    case ActionType.SET_MODAL_INVITE:
      return { ...state, setModalInvite: action.payload };
    case ActionType.SET_IS_LOGEDIN:
      return { ...state, setIsLoggingIn: action.payload };
    case ActionType.SIGN_OUT:
      return initialState;
    default:
      return state;
  }
};

export default setReducer;
