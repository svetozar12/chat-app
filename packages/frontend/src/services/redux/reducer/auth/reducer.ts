import { HYDRATE } from 'next-redux-wrapper';
import { IAuth } from './state';
import { ActionType } from '../../types';

export const initialState: IAuth = {
  remember_me: false,
  loginPrompt: false,
};
const authReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default authReducer;
