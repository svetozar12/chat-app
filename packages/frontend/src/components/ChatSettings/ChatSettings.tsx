import { css, cx } from "@emotion/css";
import React from "react";
import { useRouter } from "next/router";
import { AiOutlineUserDelete, AiOutlinePlusCircle } from "react-icons/ai";
import { Socket } from "socket.io-client";
import { getFirstChat } from "../../utils/getFirstChat";
import { useSelector, useDispatch } from "react-redux";
// services
import { IInitialSet } from "../../services/redux/reducer/setReducer/state";
import api_helper from "../../services/graphql/api_helper";
import { useCookie } from "next-cookie";
import { IAuthState } from "../../services/redux/reducer/authReducer/state";

interface IChatSettings {
  chatId: string;
}

function ChatSettings({ chatId }: IChatSettings) {
  const state = useSelector((state: { setReducer: IInitialSet }) => state.setReducer);
  const authState = useSelector((state: { authReducer: IAuthState }) => state.authReducer);

  const [users, setUsers] = React.useState<string[]>([]);
  const dispatch = useDispatch();
  const route = useRouter();
  const cookie = useCookie();
  const id = cookie.get("id") as string;
  const token = cookie.get("token") as string;

  const emitFriendRequest = async () => {
    authState.ws?.emit("friend_request");
  };
  const getMembers = async () => {
    try {
      const res = await api_helper.chatroom.getById(chatId, id, token);
      const data = res.members;

      setUsers(data);
      return true;
    } catch (error) {
      return false;
    }
  };

  const deleteMember = async (user: string) => {
    try {
      await api_helper.chatroom.update(id, chatId, token, user);
      return true;
    } catch (error) {
      return false;
    }
  };

  React.useEffect(() => {
    setUsers([]);
    getMembers();
    authState.ws?.on("inviting_multiple_users", ({ users }) => {
      setUsers((prev) => [...prev, ...users]);
    });
  }, [route.asPath]);

  const redirect = async (user: string) => {
    const updated_users = users.filter((element) => element !== user);
    setUsers(updated_users);
    if (updated_users.length === 2) {
      const redirect = await getFirstChat(id, token);

      route.push(`/${redirect._id}`);
    }
  };
  return (
    <>
      <div
        className={cx(
          "flex",
          css`
            position: relative;
            bottom: 0;
            transition: 0.3s;
            opacity: ${state.setChatSettings ? "1" : "0"};
            width: 100%;
            z-index: 11;
            height: 80vh;
            background: var(--main-white);
            border-left: 1px solid rgba(0, 0, 0, 0.1);
            flex-direction: column;
            justify-content: flex-start;
            overflow-x: hidden;
          `,
        )}
      >
        <h1
          className={css`
            color: var(--main-black);
            justify-content: center;
            white-space: nowrap;
            padding: 0 0.5rem;
            width: 70%;
            display: flex;
            justify-content: center;
            @media (max-width: 1000px) {
              font-size: 4vw;
            }
          `}
        >
          Members in chat
        </h1>

        {users.map((item, index) => {
          return (
            <div
              key={index}
              className={cx(
                "flex",
                css`
                  justify-content: space-between;
                `,
              )}
            >
              <h2
                className={cx(
                  "flex",
                  css`
                    flex-direction: column;
                    color: var(--main-black);
                    white-space: nowrap;
                  `,
                )}
              >
                {item}
              </h2>
              <AiOutlineUserDelete
                onClick={() => {
                  deleteMember(item);
                  emitFriendRequest();
                  redirect(item);
                }}
                className={css`
                  width: 2rem;
                  height: 2rem;
                  cursor: pointer;
                  &:hover {
                    color: red;
                  }
                `}
              />
            </div>
          );
        })}
        {users.length > 2 && (
          <div
            onClick={() => {
              dispatch({
                type: "SET_MODAL_INVITE",
                payload: !state.setModalInvite,
              });
            }}
            className={cx(
              "flex",
              css`
                position: relative;
                z-index: 101;
                width: 2rem;
                height: 2rem;
                cursor: pointer;
                padding: 2rem 0.5rem;
                border-radius: 5px;
                justify-content: space-between;
                width: 70%;
                height: 2rem;
                whitespace: nowrap;
                &:hover {
                  background: rgba(0, 0, 0, 0.1);
                }
              `,
            )}
          >
            <h2
              className={css`
                color: var(--main-black);
                margin: 0;
                @media (max-width: 1344px) {
                  font-size: 1.5vw;
                  word-break: keep-all;
                }
              `}
            >
              Add more users
            </h2>
            <div className="flex">
              <AiOutlinePlusCircle
                className={css`
                  width: 2rem;
                  height: 2rem;
                `}
              />
            </div>
          </div>
        )}
      </div>
    </>
  );
}

export default ChatSettings;
