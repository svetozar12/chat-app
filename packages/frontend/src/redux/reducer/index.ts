import { combineReducers } from "redux";
import reducer from "./authReducer";
import setReducer from "./setReducer";
import saveInputReducer from "./save_inputReducer";
import messageReducer from "./messageReducer";

const reducers = combineReducers({
  authReducer: reducer,
  setReducer,
  messageReducer,
  saveInputReducer,
});

export default reducers;
