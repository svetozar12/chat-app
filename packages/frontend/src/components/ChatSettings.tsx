/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from "@emotion/react";
import { css } from "@emotion/css";
import React from "react";
import axios from "axios";
import { useRouter } from "next/router";
import { requestUrl } from "../utils/hostUrl_requestUrl";
import { AiOutlineUserDelete } from "react-icons/ai";
import { Socket } from "socket.io-client";
import { getFirstChat } from "../utils/getFirstChat";
import { AiOutlinePlusCircle } from "react-icons/ai";
import { useSelector, useDispatch } from "react-redux";
import { InitialState2 } from "../redux/state";
import styled from "@emotion/styled";
const Add_users = styled.div`
  position: relative;
  z-index: 101;
  width: 2rem;
  height: 2rem;
  cursor: pointer;
  padding: 2rem 0.5rem;
  border-radius: 5px;
  justify-content: space-between;
  width: 70%;
  height: 2rem;
  whitespace: nowrap;
  &:hover {
    background: rgba(0, 0, 0, 0.1);
  }
`;

const Add_users_button = styled.h2`
  color: var(--main-black);
  margin: 0;
  @media (max-width: 1344px) {
    font-size: 1.5vw;
    word-break: keep-all;
  }
`;

const Chat_settings_members = styled.div`
  width: 70%;
  padding: 0 0.5rem;
  justify-content: space-between;
`;

const Chat_settings_nav = styled.div`
  position: relative;
  bottom: 0;
  width: 100%;
  z-index: 11;
  height: 80vh;
  background: var(--main-white);
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  flex-direction: column;
  justify-content: flex-start;
  overflow: auto;
`;

function ChatSettings({
  chatId,
  socketRef,
  setLocalStatus,
  cookieName,
}: {
  chatId: string;
  socketRef: Socket;
  cookieName: string;
  setLocalStatus: React.Dispatch<React.SetStateAction<string>>;
}) {
  const dispatch = useDispatch();
  const state = useSelector(
    (state: { setReducer: InitialState2 }) => state.setReducer,
  );

  const [users, setUsers] = React.useState<string[]>([]);
  const route = useRouter();
  const emitFriendRequest = async () => {
    socketRef?.emit("friend_request");
  };
  const getMembers = async () => {
    try {
      const res = await axios.get(`${requestUrl}/chat-room/${chatId}`);
      const data = res.data.Message[0].members;
      setUsers(data);
      return true;
    } catch (error) {
      return false;
    }
  };

  const deleteMember = async (user: string) => {
    try {
      await axios.put(`${requestUrl}/chat-room/${chatId}?username=${user}`);
      return true;
    } catch (error) {
      return false;
    }
  };

  React.useEffect(() => {
    setUsers([]);
    getMembers();
    socketRef.on("inviting_multiple_users", ({ users }) => {
      setUsers((prev) => [...prev, ...users]);
    });
  }, [route.asPath]);

  const redirect = async (user: string) => {
    const updated_users = users.filter((element) => element !== user);
    setUsers(updated_users);
    if (updated_users.length === 2) {
      const redirect = await getFirstChat(cookieName);

      route.push(`/${redirect._id}`);
    }
    setLocalStatus(user);
    setLocalStatus("");
  };
  return (
    <>
      <Chat_settings_nav className={`flex`}>
        <h1
          className={css`
            color: var(--main-black);
            justify-content: center;
            white-space: nowrap;
            padding: 0 0.5rem;
            width: 70%;
            display: flex;
            justify-content: flex-start;
            @media (max-width: 1000px) {
              font-size: 4vw;
            }
          `}
        >
          Members in chat
        </h1>
        {users.map((item, index) => {
          return (
            <Chat_settings_members key={index} className="flex">
              <h2
                style={{
                  flexDirection: "column",
                  color: "var(--main-black)",
                  whiteSpace: "nowrap",
                }}
                className="flex"
              >
                {item}
              </h2>
              <AiOutlineUserDelete
                onClick={() => {
                  deleteMember(item);
                  emitFriendRequest();
                  redirect(item);
                }}
                className={css`
                  width: 2rem;
                  height: 2rem;
                  cursor: pointer;
                  &:hover {
                    color: red;
                  }
                `}
              />
            </Chat_settings_members>
          );
        })}
        {users.length > 2 && (
          <Add_users
            onClick={() => {
              dispatch({
                type: "SET_MODAL_INVITE",
                payload: !state.setModalInvite,
              });
            }}
            className="flex"
          >
            <Add_users_button>Add more users</Add_users_button>
            <div className="flex">
              <AiOutlinePlusCircle style={{ width: "2rem", height: "2rem" }} />
            </div>
          </Add_users>
        )}
      </Chat_settings_nav>
    </>
  );
}

export default ChatSettings;
