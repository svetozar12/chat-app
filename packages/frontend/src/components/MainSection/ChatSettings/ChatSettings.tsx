import { css, cx } from '@emotion/css';
import React from 'react';
import { useRouter } from 'next/router';
import { AiOutlineUserDelete, AiOutlinePlusCircle } from 'react-icons/ai';
import { useCookie } from 'next-cookie';
import { Heading, HStack, VStack } from '@chakra-ui/react';
import generic from '../../../utils/generic';
// services
import s from './ChatSettings.module.css';
import useThemeColors from '../../../hooks/useThemeColors';
import { STATE } from 'services/redux/reducer';
import { bindActionCreators, Dispatch } from 'redux';
import { toggleInviteModal } from 'services/redux/reducer/toggles/actions';
import { connect } from 'react-redux';
import { IWebSocket } from 'services/redux/reducer/websocket/state';
import { IToggle } from 'services/redux/reducer/toggles/state';
import sdk from 'services/sdk';

interface IChatSettings {
  chatId: string;
  ws: IWebSocket;
  toggle: IToggle;
  toggleInviteModal: typeof toggleInviteModal;
}

function ChatSettings(props: IChatSettings) {
  const { chatId, ws, toggle } = props;
  const [users, setUsers] = React.useState<string[]>([]);
  const route = useRouter();
  const cookie = useCookie();
  const id: string = cookie.get('id');
  const token: string = cookie.get('token');

  const emitFriendRequest = async () => {
    ws.ws?.emit('friend_request');
  };
  const getMembers = async () => {
    try {
      const res = await sdk.chat.getChatById({ auth: { userId: id, AccessToken: token }, chat_id: chatId });
      const {
        getChatById: { members },
      } = res;
      setUsers(members);
      return true;
    } catch (error) {
      return false;
    }
  };

  const deleteMember = async (user: string) => {
    try {
      await sdk.chat.updateChat({ auth: { userId: id, AccessToken: token }, chat_id: chatId, username: user });
      return true;
    } catch (error) {
      return false;
    }
  };

  React.useEffect(() => {
    setUsers([]);
    getMembers();
    ws.ws?.on('inviting_multiple_users', ({ users }) => {
      setUsers((prev) => [...prev, ...users]);
    });
  }, [route.asPath]);

  const redirect = async (user: string) => {
    const updatedUsers = users.filter((element) => element !== user);
    setUsers(updatedUsers);
    if (updatedUsers.length === 2) {
      const redirect: { _id: string } = await generic.getFirstChat(id, token);

      route.push(`/${redirect._id}`);
    }
  };

  const {
    colors: { color },
  } = useThemeColors();

  return (
    <VStack mt={5} gap={5} pos="relative" transition="ease" w="full" opacity={toggle.toggleChatSettings ? 1 : 0}>
      <Heading w="70%" color={color} textAlign="center" whiteSpace="nowrap" fontSize={{ base: '4vw', md: '2vw' }}>
        Members in chat
      </Heading>

      {users.map((item, index) => (
        <HStack alignItems="center" key={index}>
          <Heading>{item}</Heading>
          <AiOutlineUserDelete
            onClick={() => {
              deleteMember(item);
              emitFriendRequest();
              redirect(item);
            }}
            className={s.removeUser}
          />
        </HStack>
      ))}
      {users.length > 2 && (
        <div
          onClick={() => toggleInviteModal(!toggle.toggleInvideModal)}
          className={cx(
            'flex',
            css`
              position: relative;
              z-index: 101;
              width: 2rem;
              height: 2rem;
              cursor: pointer;
              padding: 2rem 0.5rem;
              border-radius: 5px;
              justify-content: space-between;
              width: 70%;
              height: 2rem;
              whitespace: nowrap;
              &:hover {
                background: rgba(0, 0, 0, 0.1);
              }
            `,
          )}
        >
          <h2
            className={css`
              color: var(--main-black);
              margin: 0;
              @media (max-width: 1344px) {
                font-size: 1.5vw;
                word-break: keep-all;
              }
            `}
          >
            Add more users
          </h2>
          <div className="flex">
            <AiOutlinePlusCircle
              className={css`
                width: 2rem;
                height: 2rem;
              `}
            />
          </div>
        </div>
      )}
    </VStack>
  );
}

const mapStateToProps = (state: STATE) => ({
  ws: state.ws,
  toggle: state.toggle,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  toggleInviteModal: bindActionCreators(toggleInviteModal, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ChatSettings);
