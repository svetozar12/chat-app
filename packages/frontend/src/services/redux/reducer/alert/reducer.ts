import { AnyAction } from 'redux';
import { IAlert } from 'services/redux/reducer/alert/state';
import { ActionType } from '../../types';

const initialState: IAlert = {
  good: '', // good and bad stand for good alert and bad alert
  bad: '',
};

const alertReducer = (state = initialState, action: AnyAction) => {
  switch (action.type) {
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.payload };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.payload };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.payload };
    default:
      return state;
  }
};

export default alertReducer;
