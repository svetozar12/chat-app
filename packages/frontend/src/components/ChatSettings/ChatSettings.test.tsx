import ChatSettings from "../ChatSettings/ChatSettings";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { initialState as setState } from "../../redux/reducer/setReducer";
import SocketMock from "socket.io-mock";
const submit: any = jest.fn();
const mockStore = configureStore([]);
const store = mockStore({
  setReducer: setState,
});
let socket = new SocketMock();
jest.mock("next/router", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
}));

describe("Render connected React-redux page", () => {
  it("should create snapshot for <ChatSettings/>", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <ChatSettings
              cookieName={"greg"}
              socketRef={socket}
              setLocalStatus={submit}
              chatId={"321312313"}
            />
          </Provider>,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });

  it("should render <ChatSettings/>", () => {
    render(
      <Provider store={store}>
        <ChatSettings
          cookieName={"greg"}
          socketRef={socket}
          setLocalStatus={submit}
          chatId={"321312313"}
        />
      </Provider>,
    );
    const renderedComponent = screen.getByText("Members in chat");
    expect(renderedComponent).toBeInTheDocument();
  });
});
