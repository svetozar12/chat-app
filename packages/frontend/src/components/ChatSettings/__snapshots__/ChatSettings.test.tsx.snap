// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render connected React-redux page should create snapshot for <ChatSettings/> 1`] = `
<div
  className="flex css-1gszkkh"
>
  <h1
    className="css-zt1gbb"
  >
    Members in chat
  </h1>
</div>
`;
